arangod/VocBase/Methods/AqlUserFunctions.cpp

```diff
@@ -0,0 +1,445 @@
+  }
+
+  VPackSlice countSlice = queryResult.result->slice();
+  if (!countSlice.isArray() || (countSlice.length() != 1)) {
+    return Result(TRI_ERROR_INTERNAL, "bad query result for deleting AQL User Functions");
+  }
+
+  deleteCount = countSlice[0].getInt();
+  reloadAqlUserFunctions();
+  return Result();
+}
+
+Result arangodb::registerUserFunction(TRI_vocbase_t* vocbase,
+                                      velocypack::Slice userFunction,
+                                      bool& replacedExisting
+                                      ) {
+  Result res;
+  std::string name;
+  try{
+    auto vname = userFunction.get("name");
+    if (!vname.isString()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_NAME, "function name has to be provided as a string");
+    }
+    name = vname.copyString();
+    if (name.empty()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_NAME, "function name has to be provided and must not be empty");
+    }
+  }
+  catch (...) {
+    return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_NAME, "missing mandatory property - 'name'");
+  }
+
+  if (!isValidFunctionName(name)) {
+    return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_NAME,
+                  std::string("Error creating AQL User Function: '") +
+                  name +
+                  "' is not a valid name.");
+  }
+
+  std::string code;
+  std::string tmp;
+  try {
+    auto cvString = userFunction.get("code");
+    if (!cvString.isString()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "expecting function or string");
+    }
+    tmp = cvString.copyString();
+
+    if (tmp.empty()) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "expecting function - must not be empty.");
+    }
+    code = std::string("(") + tmp + "\n)";
+  }
+  catch (...) {
+      return Result(TRI_ERROR_QUERY_FUNCTION_INVALID_CODE,
+		    "missing mandatory property - 'code'");
+  }
+  
+  bool isDeterministic = false;
+  try {
+    isDeterministic = userFunction.get("isDeterministic").getBool();
+  }
+  catch (...) {}
+
+  {
+    ISOLATE;
+    bool throwV8Exception = (isolate != nullptr);
+    V8ContextDealerGuard dealerGuard(res, isolate, vocbase, true /*allowModification*/);
+    if(res.fail()){
+      return res;
+    }
+
+    std::string testCode("(function() { var callback = ");
+    testCode += tmp + "; return callback; })()";
+    v8::HandleScope scope(isolate);
+
+
+    v8::Handle<v8::Value> result;
+    {
+      v8::TryCatch tryCatch;
+
+      result = TRI_ExecuteJavaScriptString(isolate,
+                                           isolate->GetCurrentContext(),
+                                           TRI_V8_STD_STRING(isolate, testCode),
+                                           TRI_V8_ASCII_STRING(isolate, "userFunction"),
+                                           false);
+
+      int state = 0;
+      if (result.IsEmpty()) {
+        state = 1;
```

Comment:
why store the result in a variable and not simply return from here?

Reply:
it is used to create the test string below

Reply:
you can directly return from here with the error.

Reply:
all 3 conditions create different error messages. The return is done outside of that block.

Reply:
I understand. But who says the return needs to be outside of that block?